README.md
python-bayeux
==========================

A bayeux client for python.  Built on gevent and requests.

As of version 1.0.0, code using this library must do

```python
from gevent import monkey
monkey.patch_all()
```

or some other choice of patch_* functions to prepare gevent before importing this library.

Python 3 is officially supported.


Tests
-----

To run tests, install py.test and pytest-cov in your virtualenv and

$ py.test -rw -rs --cov=src/python_bayeux/ --cov-report html:coverage

View test coverage results at ``./coverage``.


Credits
-------

- [modern-package-template](http://pypi.python.org/pypi/modern-package-template)
- [requests](https://pypi.python.org/pypi/requests)
- [gevent](http://www.gevent.org/)
- [py.test](http://doc.pytest.org/en/latest/index.html)
- [pytest-cov](https://pypi.python.org/pypi/pytest-cov)


NEWS.txt
News
====

1.0.0
---

*Release date: 2021-01-08*

* Breaking change: No more official support for python 2: **If you need to run on python 2, stay on the 0.1.x series.**
* Breaking change: This library no longer calls gevent's monkey patch internally.  Code using this library must do

```
from gevent import monkey
monkey.patch_all()
```

or some other choice of patch_* functions to prepare gevent before importing this library.  See

https://github.com/gevent/gevent/issues/1231#issuecomment-393945383

for some context.

0.1.7
---

*Release date: 2018-10-07*

* Raise exceptions in main greenlet

0.1.6
---

*Release date: 2018-10-03*

* Improved error handling

0.1.5
---

*Release date: 2018-09-03*

* Callbacks now wait for resubscribe to finish

0.1.4
---

*Release date: 2018-08-29*

* If connect tells us to handshake again, resubscribe after handshake

0.1.3
---

*Release date: 2018-07-24*

* Fix unknown client error code - thanks for the report @ivanpricewaycom!

0.1.2
---

*Release date: 2018-01-04*

* Fix logging problem with unicode - thanks @christophe-gabard!

0.1
---

*Release date: 2016-08-26*

* Initial release.



tests/test_it.py
'''
    Copyright (c) 2016, Salesforce.org
    All rights reserved.

    Redistribution and use in source and binary forms, with or without
    modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.
    * Neither the name of Salesforce.org nor the names of
      its contributors may be used to endorse or promote products derived
      from this software without specific prior written permission.

    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
    FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
    COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
    POSSIBILITY OF SUCH DAMAGE.
'''

from gevent import monkey
monkey.patch_all()

from getpass import getpass
from pytest import fixture
from pytest import raises
from python_bayeux import BayeuxClient
from python_bayeux import RepeatedTimeoutException
from types import MethodType
import logging
import gevent
import simplejson as json
import random
from string import ascii_letters
import requests


class ClientOne(BayeuxClient):
    def test_shutdown(self, connect_response_element):
        self.value = connect_response_element
        self.publish(
            'Ok, I got {0}. Checking if it is right...'.format(
                self.value['data']['chat']
            ),
            'TestUser'
        )
        self.shutdown()

    def publish(self, message, username):
        super(ClientOne, self).publish(
            '/chat/demo',
            {
                'chat': message,
                'user': username
            }
        )


class ClientTwo(BayeuxClient):
    def test_print_some_stuff(self, connect_response_element):
        if not hasattr(self, 'counter'):
            self.counter = 0

        if not hasattr(self, 'detected_randoms'):
            self.detected_randoms = set([])
        try:
            self.detected_randoms.add(json.loads(
                connect_response_element['data']['chat']
            )['generated_random'])
        except json.JSONDecodeError:  # Ignore non-JSON chats
            return

        self.counter += 1
        if self.counter == 5:
            self.shutdown()

    def start_publish_loop(self):
        loop_greenlet = gevent.Greenlet(self._publish_loop)
        self.greenlets.append(loop_greenlet)
        loop_greenlet.start()

    def _publish_loop(self):
        while not self.stop_greenlets:
            self.publish(
                'I am a test looping.',
                'TestUserLooping'
            )
            gevent.sleep(1)

    def publish(self, message, username):
        if not hasattr(self, 'generated_randoms'):
            self.generated_randoms = set([])

        generated_random = random.random()

        super(ClientTwo, self).publish(
            '/chat/demo',
            {
                'chat': json.dumps({
                    'message': message,
                    'generated_random': generated_random,
                    'sending_client_id': self.client_id
                }),
                'user': username
            }
        )
        self.generated_randoms.add(generated_random)


@fixture(scope='module')
def please_start_demo():
    getpass(
        '\nPlease start the cometd-demo and hit enter when ready. '
        'This is often done by going to the cometd-demo/ directory '
        'in the cometd distribution and running \n'
        '$ mvn jetty:run\n'
    )


def test_one_client(please_start_demo):
    # Using getpass() only because I know it handles the py.test tty redirect
    # well
    code = ''.join(random.choice(ascii_letters) for i in range(12))
    getpass(
        'Point a browser to \n'
        'http://localhost:8080/jquery-examples/chat/\n'
        'Enter a chat name, and click the Join button.\n'
        'Watch for the chat prompt.  When prompted, '
        'copy the following string into the chat'
        ' : {0}\n'
        'Hit enter when you are ready:'.format(
            code
        )
    )
    with ClientOne('http://localhost:8080/cometd') as client:
        client.subscribe('/chat/demo', 'test_shutdown')
        client.publish(
            'Please enter the test code!',
            'TestUser'
        )
        try:
            client.block()
        except KeyboardInterrupt:
            assert False
        assert client.value['data']['chat'] == code


def test_one_client_go_then_block(please_start_demo):
    # Using getpass() only because I know it handles the py.test tty redirect
    # well
    code = ''.join(random.choice(ascii_letters) for i in range(12))
    getpass(
        'Point a browser to \n'
        'http://localhost:8080/jquery-examples/chat/\n'
        'Enter a chat name, and click the Join button.\n'
        'Watch for the chat prompt.  When prompted, '
        'copy the following string into the chat'
        ' : {0}\n'
        'Hit enter when you are ready:'.format(
            code
        )
    )
    with ClientOne('http://localhost:8080/cometd') as client:
        client.subscribe('/chat/demo', 'test_shutdown')
        client.publish(
            'Please enter the test code!',
            'TestUser'
        )
        client.go()
        try:
            client.block()
        except KeyboardInterrupt:
            assert False
        assert client.value['data']['chat'] == code


def test_one_client_subscribe_timeouts(caplog, please_start_demo, monkeypatch):
    caplog.set_level(logging.INFO)

    def patched_send_message(self, payload, **kwargs):
        if 'channel' in payload and payload['channel'] == '/meta/subscribe':
            raise requests.exceptions.ReadTimeout()
        else:
            return self._real_send_message(payload, **kwargs)

    with ClientOne('http://localhost:8080/cometd') as client:
        client._real_send_message = client._send_message
        client._send_message = MethodType(patched_send_message, client)

        client.subscribe('/chat/demo', 'test_shutdown')
        client.publish(
            'No need to enter a test code.  Just wait...',
            'TestUser'
        )
        with raises(RepeatedTimeoutException) as exc_info:
            client.block()
        assert exc_info.value.greenlet_name == 'subscribe'

        assert not hasattr(client, 'value')


def test_one_client_unsubscribe_timeouts(please_start_demo, monkeypatch):
    def patched_send_message(self, payload, **kwargs):
        if 'channel' in payload and payload['channel'] == '/meta/unsubscribe':
            raise requests.exceptions.ReadTimeout()
        else:
            return self._real_send_message(payload, **kwargs)

    with ClientOne('http://localhost:8080/cometd') as client:
        client._real_send_message = client._send_message
        client._send_message = MethodType(patched_send_message, client)

        client.subscribe('/chat/demo', 'test_shutdown')
        client.publish(
            'No need to enter a test code.  Just wait...',
            'TestUser'
        )
        client.unsubscribe('/chat/demo')
        with raises(RepeatedTimeoutException) as exc_info:
            client.block()
        assert exc_info.value.greenlet_name == 'unsubscribe'

        assert not hasattr(client, 'value')


def test_one_client_subscribe_timeouts_recover(please_start_demo, monkeypatch):
    # Using getpass() only because I know it handles the py.test tty redirect
    # well
    code = ''.join(random.choice(ascii_letters) for i in range(12))
    getpass(
        'Point a browser to \n'
        'http://localhost:8080/jquery-examples/chat/\n'
        'Enter a chat name, and click the Join button.\n'
        'Watch for the chat prompt.  When prompted, '
        'copy the following string into the chat'
        ' : {0}\n'
        'Hit enter when you are ready:'.format(
            code
        )
    )

    def patched_send_message(self, payload, **kwargs):
        if 'channel' in payload and payload['channel'] == '/meta/subscribe':
            self.test_num_timeouts = 0  \
                if not hasattr(self, 'test_num_timeouts') \
                else self.test_num_timeouts + 1

            if self.test_num_timeouts > 5:
                return self._real_send_message(payload, **kwargs)

            raise requests.exceptions.ReadTimeout()
        else:
            return self._real_send_message(payload, **kwargs)

    with ClientOne('http://localhost:8080/cometd') as client:
        client._real_send_message = client._send_message
        client._send_message = MethodType(patched_send_message, client)

        client.subscribe('/chat/demo', 'test_shutdown')
        client.publish(
            'In about 30 seconds, please enter the test code!',
            'TestUser'
        )
        try:
            client.block()
        except KeyboardInterrupt:
            assert False
        assert client.value['data']['chat'] == code


def test_one_client_unsubscribe_timeouts_recover(please_start_demo,
                                                 monkeypatch):
    # Using getpass() only because I know it handles the py.test tty redirect
    # well
    code = ''.join(random.choice(ascii_letters) for i in range(12))
    getpass(
        'Point a browser to \n'
        'http://localhost:8080/jquery-examples/chat/\n'
        'Enter a chat name, and click the Join button.\n'
        'Watch for the chat prompt.  When prompted, '
        'copy the following string into the chat'
        ' : {0}\n'
        'Hit enter when you are ready:'.format(
            code
        )
    )

    def patched_send_message(self, payload, **kwargs):
        if 'channel' in payload and payload['channel'] == '/meta/unsubscribe':
            self.test_num_timeouts = 0  \
                if not hasattr(self, 'test_num_timeouts') \
                else self.test_num_timeouts + 1

            if self.test_num_timeouts > 5:
                return self._real_send_message(payload, **kwargs)

            raise requests.exceptions.ReadTimeout()
        else:
            return self._real_send_message(payload, **kwargs)

    with ClientOne('http://localhost:8080/cometd') as client:
        client._real_send_message = client._send_message
        client._send_message = MethodType(patched_send_message, client)

        client.subscribe('/chat/demo', 'test_shutdown')
        client.publish(
            'Please wait...',
            'TestUser'
        )
        client.unsubscribe('/chat/demo')
        client.publish(
            'Please keep waiting...',
            'TestUser'
        )
        client.subscribe('/chat/demo', 'test_shutdown')
        client.publish(
            'Please enter the test code!',
            'TestUser'
        )
        try:
            client.block()
        except KeyboardInterrupt:
            assert False
        assert client.value['data']['chat'] == code


def test_two_clients(please_start_demo):
    with ClientTwo('http://localhost:8080/cometd') as client1:
        client1.go()

        with ClientTwo('http://localhost:8080/cometd') as client2:
            client2.subscribe('/chat/demo', 'test_print_some_stuff')

            client2.publish(
                'I am a test 2.',
                'TestUser2'
            )

            client1.start_publish_loop()
            try:
                client2.block()
            except KeyboardInterrupt:
                assert False
            for element in client2.detected_randoms:
                assert element in client1.generated_randoms


class BrokenClient(BayeuxClient):
    def test_exception(self, connect_response_element):
        raise Exception('ouch')

    def publish(self, message, username):
        super(BrokenClient, self).publish(
            '/chat/demo',
            {
                'chat': message,
                'user': username
            }
        )


def test_broken_client(please_start_demo):
    # Using getpass() only because I know it handles the py.test tty redirect
    # well
    getpass(
        'Point a browser to \n'
        'http://localhost:8080/jquery-examples/chat/\n'
        'Enter a chat name, and click the Join button.\n'
        'Hit enter when you are ready:'
    )

    with BrokenClient('http://localhost:8080/cometd') as client:
        client.subscribe('/chat/demo', 'test_exception')

        client.publish(
            'Please enter any chat message.',
            'TestUser'
        )

        client.go()
        with raises(Exception) as e:
            client.block()
            assert str(e) == 'ouch'


def test_broken_client_same_greenlet(please_start_demo):
    # Using getpass() only because I know it handles the py.test tty redirect
    # well
    getpass(
        'Point a browser to \n'
        'http://localhost:8080/jquery-examples/chat/\n'
        'Enter a chat name, and click the Join button.\n'
        'Hit enter when you are ready:'
    )

    with BrokenClient('http://localhost:8080/cometd') as client:
        client.subscribe('/chat/demo', 'test_exception')

        client.publish(
            'Please enter any chat message.',
            'TestUser'
        )

        with raises(Exception) as e:
            client.block()
            assert str(e) == 'ouch'


def test_broken_client_local_loop(please_start_demo):
    # Using getpass() only because I know it handles the py.test tty redirect
    # well
    getpass(
        'Point a browser to \n'
        'http://localhost:8080/jquery-examples/chat/\n'
        'Enter a chat name, and click the Join button.\n'
        'Hit enter when you are ready:'
    )

    with BrokenClient('http://localhost:8080/cometd') as client:
        client.subscribe('/chat/demo', 'test_exception')

        client.publish(
            'Please enter any chat message.',
            'TestUser'
        )

        client.go()
        try:
            while client.exception is None:
                gevent.sleep(0.5)
        except KeyboardInterrupt:
            assert False

        assert isinstance(
            client.exception,
            Exception
        )

        assert str(client.exception) == 'ouch'


src/python_bayeux/__init__.py
'''
    Copyright (c) 2016, Salesforce.org
    All rights reserved.

    Redistribution and use in source and binary forms, with or without
    modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.
    * Neither the name of Salesforce.org nor the names of
      its contributors may be used to endorse or promote products derived
      from this software without specific prior written permission.

    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
    FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
    COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
    POSSIBILITY OF SUCH DAMAGE.
'''

import simplejson as json
import gevent
import gevent.queue
import requests
import requests.exceptions
from datetime import datetime
from copy import deepcopy

import logging
LOG = logging.getLogger('python_bayeux')

# See https://docs.cometd.org/current/reference/#_bayeux for bayeux reference


class BayeuxClient(object):
    def __init__(self, endpoint=None, oauth_session=None, start=True):
        self.endpoint = endpoint
        self.oauth_session = \
            requests.Session() if oauth_session is None else oauth_session
        self.shutdown_called = False
        self.shutdown_completed = False

        # Inbound
        self.message_queue = gevent.queue.Queue()

        # Outbound
        self.subscription_queue = gevent.queue.Queue()
        self.unsubscription_queue = gevent.queue.Queue()
        self.publication_queue = gevent.queue.Queue()

        self.channel_ids = {}

        self.subscription_callbacks = {}

        # handshake() has a side effect of initializing self.message_counter
        self.handshake()

        self.connect_timeout = None
        self.disconnect_complete = False
        self.executing = False
        self.stop_greenlets = False
        self.waiting_for_resubscribe = False
        self.go_called = False
        self.exception = None

        self.outbound_greenlets = []
        for method in (self._subscribe_greenlet,
                       self._unsubscribe_greenlet,
                       self._publish_greenlet):
            new_greenlet = gevent.Greenlet(method)
            new_greenlet.link_exception(self._exception_callback)
            self.outbound_greenlets.append(new_greenlet)

        self.inbound_greenlets = []
        connect_greenlet = gevent.Greenlet(self._connect_greenlet)
        connect_greenlet.link_exception(self._exception_callback)
        self.inbound_greenlets.append(connect_greenlet)

        self.greenlets = self.outbound_greenlets + self.inbound_greenlets

        if start:
            self.start()

    def handshake(self, **kwargs):
        self.message_counter = 1

        handshake_payload = {
            # MUST
            'channel': '/meta/handshake',
            'supportedConnectionTypes': ['long-polling'],
            'version': '1.0',
            # MAY
            'id': None,
            'minimumVersion': '1.0'
        }
        handshake_payload.update(kwargs)
        handshake_response = self._send_message(handshake_payload)

        # TODO: No error checking here
        self.client_id = handshake_response[0]['clientId']

        # Connect one time to get the server's timeout advive
        initial_connect_response = self.connect(initial=True)
        initial_connect_response_payload = initial_connect_response[0]

        # TODO: handle other 'reconnect' values
        if initial_connect_response_payload['successful']:
            # Convert to seconds
            if 'advice' in initial_connect_response_payload:
                advice = initial_connect_response_payload['advice']
                self.connect_timeout = advice['timeout'] / 1000.0
        # TODO: if not successful, then what?

    def disconnect(self):
        disconnect_response = self._send_message({
            # MUST
            'channel': '/meta/disconnect',
            'clientId': None,
            # MAY
            'id': None
        })
        self.disconnect_complete = True
        return disconnect_response

    def connect(self, initial=False):
        connect_request_payload = {
            # MUST
            'channel': '/meta/connect',
            'connectionType': 'long-polling',
            'clientId': None,
            # MAY
            'id': None
        }

        timeout = None if initial else self.connect_timeout

        connect_response = self._send_message(
            connect_request_payload,
            timeout=timeout
        )

        return connect_response

    def _send_message(self, payload, **kwargs):
        if 'id' in payload:
            payload['id'] = str(self.message_counter)
            self.message_counter += 1

        if 'clientId' in payload:
            payload['clientId'] = self.client_id

        LOG.info('_send_message(): payload: {0}  kwargs: {1}'.format(
            str(payload),
            str(kwargs)
        ))

        response = self.oauth_session.post(
            self.endpoint,
            data=json.dumps(payload),
            **kwargs
        )

        LOG.info(
            u'_send_message(): response status code: {0}  '
            u'response.text: {1}'.format(
                response.status_code,
                response.text
            )
        )

        if len(response.text) == 0:
            return ''

        return response.json()

    def _connect_greenlet(self):
        connect_response = None

        while not self.stop_greenlets:
            try:
                connect_response = self.connect()
            except requests.exceptions.ReadTimeout:
                LOG.info(
                    'connect greenlet timed out {0}'.format(
                        datetime.now()
                    )
                )
            else:
                if not isinstance(connect_response, list):
                    raise UnexpectedConnectResponseException(
                        str(connect_response)
                    )

                messages = []
                handshake_required = False
                for element in connect_response:
                    channel = element['channel']

                    if channel == '/meta/connect':
                        if not element['successful'] and \
                           element['error'] == '403::Unknown client':

                            # TODO: support handshake advice interval
                            if element['advice']['reconnect'] == 'handshake':
                                handshake_required = True
                    else:
                        # We got a push!
                        messages.append(element)

                if len(messages) > 0:
                    self.message_queue.put(messages)

                if handshake_required:
                    self.handshake()
                    self._resubscribe()

    def _execute_greenlet(self):
        self.executing = True
        channel = None
        while True:
            try:
                message_queue_messages = self.message_queue.get(
                    timeout=1
                )
            except gevent.queue.Empty:
                if self.stop_greenlets:
                    LOG.info(
                        'execute greenlet is stopping: '
                        'client id {0} at {1}'.format(
                            self.client_id,
                            str(datetime.now())
                        )
                    )
                    break
                else:
                    LOG.info(
                        'execute greenlet is NOT stopping: '
                        'client id {0} at {1}'.format(
                            self.client_id,
                            str(datetime.now())
                        )
                    )
                    continue

            LOG.info('client id {0} found message info {1} at {2}'.format(
                self.client_id,
                message_queue_messages,
                datetime.now()
            ))

            while self.waiting_for_resubscribe:
                gevent.sleep(0.5)

            for message_queue_message in message_queue_messages:
                channel = message_queue_message['channel']
                for callback in self.subscription_callbacks[channel]:
                    getattr(self, callback)(message_queue_message)

    def subscribe(self, channel, callback=None, **kwargs):
        LOG.info('enqueueing subscription for channel {0}'.format(
            channel
        ))
        subscription_queue_message = {
            'channel': channel
        }
        subscription_queue_message.update(kwargs)

        if channel not in self.subscription_callbacks:
            self.subscription_callbacks[channel] = []
            self.subscription_queue.put(subscription_queue_message)

        self.subscription_callbacks[channel].append(callback)

    def _resubscribe(self):
        self.waiting_for_resubscribe = True

        current_subscriptions = deepcopy(self.subscription_callbacks)
        self.subscription_callbacks.clear()
        for channel, callbacks in current_subscriptions.items():
            for callback in callbacks:
                self.subscribe(channel, callback)

        self.waiting_for_resubscribe = False

    def _subscribe_greenlet(self, successive_timeout_threshold=20,
                            timeout_wait=5):
        channel = None

        successive_timeouts = 0
        while True:
            try:
                subscription_queue_message = self.subscription_queue.get(
                    timeout=1
                )
                channel = subscription_queue_message['channel']
            except gevent.queue.Empty:
                if self.stop_greenlets:
                    break
                else:
                    continue

            subscribe_request_payload = {
                # MUST
                'channel': '/meta/subscribe',
                'subscription': channel,
                'clientId': None,
                # MAY
                'id': None
            }

            subscribe_responses = []
            try:
                subscribe_responses = self._send_message(
                    subscribe_request_payload
                )
            except requests.exceptions.ReadTimeout:
                successive_timeouts += 1

                if successive_timeouts > successive_timeout_threshold:
                    raise RepeatedTimeoutException('subscribe')

                gevent.sleep(timeout_wait)
                self.subscription_queue.put(subscription_queue_message)
            else:
                successive_timeouts = 0

            for element in subscribe_responses:
                if not element['successful'] and \
                   element['error'] == '403::Unknown client':
                    # Just try again, and eventually connect() will re-try a
                    # handshake
                    self.subscription_queue.put(subscription_queue_message)

    def unsubscribe(self, subscription):
        LOG.info('enqueueing unsubscription for channel {0}'.format(
            subscription
        ))
        self.unsubscription_queue.put(subscription)

    def _unsubscribe_greenlet(self, successive_timeout_threshold=20,
                              timeout_wait=5):
        successive_timeouts = 0
        while True:
            unsubscription = None
            try:
                unsubscription = self.unsubscription_queue.get(timeout=1)
            except gevent.queue.Empty:
                if self.stop_greenlets:
                    break
                else:
                    continue

            unsubscribe_request_payload = {
                # MUST
                'channel': '/meta/unsubscribe',
                'subscription': unsubscription,
                'clientId': None,
                # MAY
                'id': None
            }

            try:
                self._send_message(unsubscribe_request_payload)
            except requests.exceptions.ReadTimeout:
                successive_timeouts += 1

                if successive_timeouts > successive_timeout_threshold:
                    raise RepeatedTimeoutException('unsubscribe')

                gevent.sleep(timeout_wait)
                self.unsubscription_queue.put(unsubscription)
            else:
                successive_timeouts = 0

    def publish(self, channel, payload):
        self.publication_queue.put({
            'channel': channel,
            'payload': payload
        })

    def _publish_greenlet(self):
        while True:
            publication = None
            try:
                publication = self.publication_queue.get(timeout=1)
            except gevent.queue.Empty:
                if self.stop_greenlets:
                    break
                else:
                    continue

            channel = publication['channel']

            # Note that this return value isn't going anywhere.
            publish_request_payload = {
                # MUST
                'channel': channel,
                'data': publication['payload'],
                # MAY
                'clientId': None,
                'id': None
            }

            # Directly raise exceptions
            publish_response = self._send_message(publish_request_payload)

            LOG.info('publish response: {0}'.format(str(publish_response)))

    def start(self):
        for greenlet in self.outbound_greenlets:
            greenlet.start()
        for greenlet in self.inbound_greenlets:
            greenlet.start()

    # This is how a client can be given its own execute greenlet, but not
    # block the main greenlet
    def go(self):
        if self.go_called:
            return

        self.go_called = True
        block_greenlet = gevent.Greenlet(self.block)
        block_greenlet.link_exception(self._exception_callback)
        self.greenlets.append(block_greenlet)
        block_greenlet.start()
        # give the execute greenlet a chance to start, so self.executing is
        # True if we call block() later
        gevent.sleep(0.1)

    def block(self):
        if not self.executing:
            self._execute_greenlet()
        else:
            # block the main greenlet
            while self.exception is None and any(self.greenlets):
                gevent.sleep(1)

        if self.exception is not None:
            raise self.exception

    def shutdown(self):
        if not self.shutdown_called:
            self.shutdown_called = True

            self.stop_greenlets = True

            LOG.info('client id {0} is shutting down'.format(self.client_id))

            # disconnect() runs in the main greenlet, so we want to give the
            # others a chance to finish
            # If we have been called by a callback (that is, the client wants
            # to shut down itself), then we don't want to wait for the execute
            # greenlet to stop, because we'll deadlock.
            # This also means that subclasses should not change self.greenlets
            relevant_greenlets = \
                self.greenlets[:-1] \
                if gevent.getcurrent() == self.greenlets[-1] \
                else self.greenlets

            gevent.joinall(relevant_greenlets)
            self.disconnect()
            self.shutdown_completed = True

    def _exception_callback(self, failed_greenlet):
        LOG.info(
            'client id {0} has an unhandled exception '
            'in greenlet {1}: {2}'.format(
                self.client_id,
                failed_greenlet.name,
                str(failed_greenlet.exception)
            )
        )

        self.exception = failed_greenlet.exception

        try:
            self.shutdown()
        except Exception:
            # We prefer to shut down, but if something doesn't work (like the
            # org is sending back a total requests limit error), then don't
            # worry
            pass

    def __enter__(self):
        return self

    def __exit__(self, exception_type, exception_value, traceback):
        self.shutdown()
        while not self.shutdown_completed:
            gevent.sleep(0.5)


class RepeatedTimeoutException(Exception):
    def __init__(self, greenlet_name):
        self.greenlet_name = greenlet_name
        super(RepeatedTimeoutException, self).__init__(
            'Too many timeouts'
        )


class UnexpectedConnectResponseException(Exception):
    def __init__(self, message):
        super(UnexpectedConnectResponseException, self).__init__(message)
